persist/fs/index_lookup.go

```diff
@@ -0,0 +1,200 @@
+//            1. Is closest to the desired series in the index file
+//            2. Is BEFORE the desired series in the index file (because we
+//               we scan the index file sequentially in a forward-moving manner)
+// In other words, the returned offset can always be used as a starting point to
+// begin scanning the index file for the desired series.
+func (il *nearestIndexOffsetLookup) getNearestIndexFileOffset(id ts.ID) (int64, error) {
+	idBytes := id.Data().Get()
+
+	min := 0
+	max := len(il.summaryIDsOffsets) - 1
+
+	// The summaries file only contains a fraction of the series that are in
+	// the index file itself. Because of that, the binary search that we're
+	// performing is "optimistic". We're trying to find either an exact match,
+	// OR the nearest match that is to the left of the series we're searching
+	// for (so we keep track of it everytime we move right). We start with an
+	// assumption that the best match so far is at index 0, because in the worst
+	// case scenario if we don't find a single "match", then the caller should
+	// start at index 0 and scan until they encounter an entry that tells them
+	// that the ID they're looking for does not exist (because the IDs in the
+	// index are sorted).
+	bestMatchSoFar := int64(0)
+
+	for {
+		if min > max {
+			return bestMatchSoFar, nil
+		}
+
+		idx := (max + min) / 2
+		summaryBytesMetadata := il.summaryIDsOffsets[idx]
+		compBytes := summaryBytesMetadata.ID(il.summariesMmap)
+		comparison := bytes.Compare(idBytes, compBytes)
+
+		// Found it
+		if comparison == 0 {
+			indexOffset, err := summaryBytesMetadata.IndexOffset(
+				il.summariesMmap, il.decoderStream, il.msgpackDecoder)
+			// Should never happen, either something is really wrong with the code or
+			// the file on disk was corrupted
+			if err != nil {
+				return -1, err
+			}
+			return indexOffset, nil
+		}
+
+		// idBytes is smaller than compBytes, go left
+		if comparison == -1 {
+			max = idx - 1
+			continue
+		}
+
+		// idBytes is larger than compBytes, go right
+		if comparison == 1 {
+			min = idx + 1
+			indexOffset, err := summaryBytesMetadata.IndexOffset(
+				il.summariesMmap, il.decoderStream, il.msgpackDecoder)
+			if err != nil {
+				return -1, err
+			}
+			// update the bestMatchSoFar everytime we move right
+			bestMatchSoFar = indexOffset
+			continue
+		}
+	}
+}
+
+func (il *nearestIndexOffsetLookup) close() error {
+	return munmap(il.summariesMmap)
+}
+
+// newNearestIndexOffsetLookupFromSummariesFile creates an nearestIndexOffsetLookup
+// from an index summaries file by reading the summaries file into an anonymous
+// mmap'd region, and also creating the slice of summaries offsets which is
+// required to binary search the data structure. It will also make sure that
+// the summaries file is sorted (which it always should be).
+func newNearestIndexOffsetLookupFromSummariesFile(
+	summariesFd *os.File,
+	summariesFdWithDigest digest.FdWithDigestReader,
```

Comment:
Hm, why pass both of these?  Perhaps just the second arg and expect that it's already been reset?

You can get the `*os.File` ref back with `summariesFdWithDigest.Fd()`

Reply:
I'm also having a hard time working out whether this should be a standalone method or a method on the seeker itself?  It seems a lot of the variables have lifecycles related to the seeker so it would make more sense there, but that's just off the vibe I get reading it.

Reply:
I didn't know about the Fd() method, condensed it to a single argument.

I think its a little cleaner as a stand-alone method (I did the same thing for the bloom filter) and could be potentially re-used in the reader if we ever wanted to